SQL: integer parameter validation in string functions

bpintea opened the pull request:
This PR adds integer parameter validation for the string functions taking one: LOCATE, LEFT, INSERT, REPEAT, RIGHT, SUBSTRING.

Closes #58923.
Supersedes #60281.

Co-authored-by:  yinanwu <yinanwu@tencent.com>

Commit:
SQL: integer parameter validation in string functions (#58923)

In insert, locate, substring function, when argument `start` or `length` is greater than Integer.MAX_INT OR less then Integer.MIN_INT + 1 (note that `start` need to minus 1), it causes overflow and leads to unexpected results.

Commit:
SQL: IN operator incorrectly handling specified date format (#58932)

Extract the checks into methods Check.isNumberOutOfRange and add unit tests to the three string function.

Commit:
SQL: IN operator incorrectly handling specified date format (#58932)

Update the Check.isFixedNumberAndInRange function to comply with code specifications.

Commit:
SQL: IN operator incorrectly handling specified date format (#58932)

Update the unit tests to fit new check error messages.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into fix/str_func_param_validation

bpintea committed:
Add range checks for BinaryStringNumericProcessors

- Add range checks for Left, Right, Repeat.
- Minor refactorings on initial PR changes.

elasticmachine committed:
Merge branch 'master' into fix/str_func_param_validation

elasticmachine commented:
Pinging @elastic/es-ql (:Query Languages/SQL)

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/InsertProcessorTests.java:
```diff
@@ -52,33 +52,53 @@ public void testInsertWithEdgeCases() {
-        assertEquals("A number is required; received [z]", siae.getMessage());
+        assertEquals("A fixed point number is required for [length]; received [java.lang.Character]", siae.getMessage());
+
+        assertEquals("baroobar", new Insert(EMPTY, l("foobar"), l(-2147483647), l(1), l("bar")).makePipe().asProcessor().process(null));
```
minor: Could you use Integer.MIN/MAX_VALUE to make it more clear?

wuyinan0126 replied:
@matriv OK, I will fix it

bpintea replied:
[Fix](https://github.com/elastic/elasticsearch/pull/60281/commits/7d8ecfc8aece028a965c1fad017c7e137696989c) ported.

matriv reviewed (approved):
LGTM. Nice work with the tests! (left minor comment)

PS: It would be great if in the future we could test all those invalid args at one place, maybe by iterating over the subclasses of the superclass of those processors, and they had as package private their accepted limits so they could be used in that test.

bpintea committed:
Adress review comments

Port
https://github.com/elastic/elasticsearch/pull/60281/commits/7d8ecfc8aece028a965c1fad017c7e137696989c

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix/str_func_param_validation

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/util/Check.java:
```diff
@@ -36,4 +36,15 @@ public static void notNull(Object object, String message, Object... values) {
+        }
+        Long longValue = ((Number) object).longValue();
+        if (longValue < from || longValue > to) {
+            throw new SqlIllegalArgumentException("[{}] equals [{}], out of the allowed range [{}..{}]", objectName, longValue, from, to);
```
Unless we use this format elsewhere [{} .. {}] (with two dots), I would use the comma format [{}, {}].

astefan replied:
Also, I think the style of the error message we use is first state the problem, then the actual value. Something like "[start] out of the allowed range [ ], received [{}]". Found one example (maybe there are more with out of range values) in [the Analyzer](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java#L529).

bpintea replied:
Good point, updated it.

astefan reviewed (approved):
LGTM in general, left couple of comments regarding the wording of the error message.

bpintea committed:
Adress review comments

Update one error message.

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix/str_func_param_validation
